typing/includecore.ml

```diff
@@ -374,9 +374,15 @@ let compare_records_with_representation ~loc env params1 params2 n
   =
   match compare_records ~loc env params1 params2 n labels1 labels2 with
   | None when rep1 <> rep2 ->
-      let pos = if rep2 = Record_float then Second else First in
-      Some (Unboxed_float_representation pos)
-  | err -> err
+      Some (match rep1, rep2 with
+            | _, Record_unboxed _ -> Unboxed_representation Second
+            | Record_unboxed _, _ -> Unboxed_representation First
+            | _, Record_float ->
+               Record_mismatch (Unboxed_float_representation Second)
+            | _, _ ->
+               Record_mismatch (Unboxed_float_representation First))
+  | None -> None
+  | Some err -> Some (Record_mismatch err)
```

Comment:
Idle note: this need not be part of this PR, but this code would be nicer with `(unit, ...) result` instead of `... option`.